File: backend/src/main/java/com/nimuairy/auth/service/UserService.java
package com.nimuairy.auth.service;

import com.nimuairy.auth.domain.User;
import com.nimuairy.auth.exception.domain.EmailExistException;
import com.nimuairy.auth.exception.domain.EmailNotFoundException;
import com.nimuairy.auth.exception.domain.UserNotFoundException;
import com.nimuairy.auth.exception.domain.UsernameExistException;

import java.io.IOException;
import java.util.List;
import javax.mail.MessagingException;

import org.springframework.web.multipart.MultipartFile;

public interface UserService {

    User addNewUser(String firstName, String lastName, String username, String email, String role, boolean isNonLocked, boolean isActive, MultipartFile profileImage) throws UserNotFoundException, UsernameExistException, EmailExistException, IOException;

    void deleteUser(Long id);

    User findUserByEmail(String email);

    User findUserByUsername(String username);

    List<User> getUsers();

    User register(String firstName, String lastName, String username, String email) throws UserNotFoundException, UsernameExistException, EmailExistException, MessagingException;

    void resetPassword(String email) throws EmailNotFoundException, MessagingException;

    User updateProfileImage(String username, MultipartFile profileImage) throws UserNotFoundException, UsernameExistException, EmailExistException, IOException;

    User updateUser(String currentUsername, String newFirstName, String newLastName, String newUsername, String newEmail, String role, boolean isNonLocked, boolean isActive, MultipartFile profileImage) throws UserNotFoundException, UsernameExistException, EmailExistException, IOException;

}

File: frontend/src/app/auth/service/user.service.ts
import { HttpClient, HttpErrorResponse, HttpEvent, HttpResponse } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

import { BrowserStorageService } from 'src/app/util/browser-storage.service';
import { environment } from 'src/environments/environment';
import { CustomHttpRespone } from '../model/custom-http-response';
import { User } from '../model/user';

@Injectable({
  providedIn: 'root'
})
export class UserService {
  private host: string = environment.apiUrl;

  constructor(private browserStorage: BrowserStorageService,
              private http: HttpClient) { }

  addUser(formData: FormData) {
    return this.http.post<User>(`${this.host}/user`, formData);
  }

  addUsersToLocalCache(users: User[]) {
    this.browserStorage.set('users', JSON.stringify(users));
  }

  createUserFormData(loggedInUsername: string, user: User, profileImage: File) {
    const formData = new FormData();
    formData.append('currentUsername', loggedInUsername);
    formData.append('firstName', user.firstName);
    formData.append('lastName', user.lastName);
    formData.append('username', user.username);
    formData.append('email', user.email);
    formData.append('role', user.role);
    formData.append('profileImage', profileImage);
    formData.append('isActive', JSON.stringify(user.active));
    formData.append('isNonLocked', JSON.stringify(user.notLocked));

    return formData;
  }

  deleteUser(userId: number | string) {
    return this.http.delete<CustomHttpRespone>(`${this.host}/user/${userId}`);
  }

  getUsers() {
    return this.http.get<User[]>(`${this.host}/user`);
  }

  getUsersFromLocalCache(): User[] {
    if (this.browserStorage.get('users')) {
      return JSON.parse(this.browserStorage.get('users'));
    }
    return null;
  }

  resetPassword(email: string) {
    this.http.get<CustomHttpRespone>(`${this.host}/user/resetpassword/${email}`);
  }

  updateProfileImage(formData: FormData): Observable<HttpEvent<User>> {
    return this.http.put<User>(`${this.host}/user/update-profile-image`, formData,
    {
      reportProgress: true,
      observe: 'events'
    });
  }

  updateUser(formData: FormData) {
    return this.http.patch<User>(`${this.host}/user`, formData);
  }
}
